Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -48,6 +48,7 @@ public class ProgramBuilder {
     ///
     /// The builder instance can not be used further after calling this function.
     public func finish() -> Program {
+        program.normalize()
```

Comment:
I'm not sure this really fixes the issue. 
I think the main problem with minimization is that it relies on generating programs that are invalid in the new sense (they contain holes in variables), but otherwise correct. As such, this test: https://github.com/googleprojectzero/fuzzilli/blob/5aa6f4ef3724f197a0fc55ffc9b8caf8c809ad56/Sources/Fuzzilli/Minimization/Reducer.swift#L35 should now faill basically all the time, resulting in programs not being minimized. You might see that the average program size keeps increasing or stays very large.
I think you also can't just do `normalize` before that line in the reducer, since some programs will be invalid in other ways, thus probably causing normalize to crash (I think). 
Maybe the correct fix is to add a parameter to the `check` function: `checkForVariableHoles: bool = true` which is true by default but turned off in the Reducer with a comment like `// Due to the way the reducers work, they will produce otherwise valid programs, but with variable holes, so we don't check for those. The holes will be removed after minimization is done`.

Reply:
Thanks for this suggestion, I think this would be the optimal approach and I'll add it to the PR. From what I gather we would need to call `check` with the `checkForVariableHoles` disabled in two places: one in `Reducer.swift` as you highlighted above and the other in `InliningReducer.swift`:

https://github.com/googleprojectzero/fuzzilli/blob/e26e7f78f4b364f5431547f68a412d558e56b07a/Sources/Fuzzilli/Minimization/InliningReducer.swift#L176

Reply:
So disabling the hasHoles check InliningReducer as I mentioned above, turns out, isn't sufficient. Before the assert is called, there's a call to `ProgramBuilder.finish()` which in-turn calls `Program.check()` before it returns.

https://github.com/googleprojectzero/fuzzilli/blob/e26e7f78f4b364f5431547f68a412d558e56b07a/Sources/Fuzzilli/Minimization/InliningReducer.swift#L175-L176 

https://github.com/googleprojectzero/fuzzilli/blob/e26e7f78f4b364f5431547f68a412d558e56b07a/Sources/Fuzzilli/Core/ProgramBuilder.swift#L50-L55

I'm not sure if disabling the hasHoles check in `ProgramBuilder.finish()` would be the right approach here since it's being referenced by several other components. Thoughts?

Reply:
Cool, I think the InliningReducer should be fixed now: https://github.com/googleprojectzero/fuzzilli/commit/b8e1778ca24d4bfa8aae5932c6f4a04385cf9958#diff-00251930897a0e17dddc9fdd75f5acf2

With that, does it work now when only setting checkForVariableHoles=false in the ReductionVerifier?

Reply:
Brilliant, I think this would now work. Let me run a few sanity checks and update this PR. I'll try to get this done in the next hour or so :)

Reply:
I've updated the PR to have the reducers ignore holes in the variable Map. I've also reverted the changes to ProgramBuilder since they are no longer necessary with your recent updates to the InliningReducer.